### PR TITLE
Fix two edge cases in handling of counterparty revoked commitment txn

### PR DESCRIPTION
**Pull request by TheBlueMatt**
~Based on #1481 as the tests are only useful with it.~

Discovered while working on #1466.

This adds two new commits which correct some edge-case handling of revoked counterparty commitment transactions. They're both super duper edge, so not worth rushing too much to fix them, but will be needed for #1466 so would be nice to get them soon-ish.

The first case is only relevant for users disconnecting blocks with `transaction_unconfirmed`, when the counterparty broadcasts a revoked commitment transaction, which is then reorg'd out and replaced with a non-revoked commitment transaction.

The second case handles resolving HTLCs even though we haven't yet spent them, though we have broadcast the spend transactions and should spend them soon. This, too, is only relevant in case the revoked commitment transaction gets reorg'd out and replaced with a non-revoked one (and then the HTLCs claimed by our counterparty).

### CODE REVIEW

**Review by wpaulino — COMMENTED** [review]
LGTM after rebase, nice catch and tests!

File: lightning/src/chain/channelmonitor.rs [hunk 1]
```diff
@@ -2107,7 +2124,8 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.counterparty_commitment_txn_on_chain.insert(commitment_txid, commitment_number);
 
 			log_info!(logger, "Got broadcast of non-revoked counterparty commitment transaction {}", commitment_txid);
-			fail_unbroadcast_htlcs!(self, "counterparty", height, per_commitment_data.iter().map(|(a, b)| (a, b.as_ref().map(|b| b.as_ref()))), logger);
+			fail_unbroadcast_htlcs!(self, "counterparty", commitment_txid, height,
+				per_commitment_data.iter().map(|(a, b)| (a, b.as_ref().map(|b| b.as_ref()))), logger);
```
Comment (tnull) [inline lightning/src/chain/channelmonitor.rs:2128]: Since you already started it, you could also rename these tuple decomp. variables.

File: lightning/src/chain/channelmonitor.rs [hunk 2]
```diff
@@ -2091,7 +2101,14 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 				}
 				self.counterparty_commitment_txn_on_chain.insert(commitment_txid, commitment_number);
 
-				fail_unbroadcast_htlcs!(self, "revoked counterparty", height, [].iter().map(|a| *a), logger);
+				if let Some(per_commitment_data) = per_commitment_option {
+					fail_unbroadcast_htlcs!(self, "revoked_counterparty", commitment_txid, height,
+						per_commitment_data.iter().map(|(a, b)| (a, b.as_ref().map(|b| b.as_ref()))), logger);
```
Comment (tnull) [inline lightning/src/chain/channelmonitor.rs:2106]: Same here.

File: lightning/src/chain/channelmonitor.rs [hunk 3]
```diff
@@ -2091,7 +2101,16 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 				}
 				self.counterparty_commitment_txn_on_chain.insert(commitment_txid, commitment_number);
 
-				fail_unbroadcast_htlcs!(self, "revoked counterparty", height, [].iter().map(|a| *a), logger);
+				if let Some(per_commitment_data) = per_commitment_option {
+					fail_unbroadcast_htlcs!(self, "revoked_counterparty", commitment_txid, height,
+						per_commitment_data.iter().map(|(htlc_htlc_source, b)|
```
Comment (tnull) [inline lightning/src/chain/channelmonitor.rs:2106]:
```suggestion
						per_commitment_data.iter().map(|(htlc, htlc_source)|
```

**Comment by tnull** [comment]
LGTM, feel free to squash.

**Review by tnull — DISMISSED** [review]

File: lightning/src/ln/monitor_tests.rs [hunk 4]
```diff
@@ -94,6 +94,55 @@ fn test_spendable_output<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, spendable_t
 	} else { panic!(); }
 }
 
+#[test]
+fn revoked_output_htlc_resolution_timing() {
+	// Tests that the resoluting time of HTLCs which were present in a broadcasted remote revoked
```
Comment (valentinewallace) [inline lightning/src/ln/monitor_tests.rs:99]: nit: s/the resoluting time of //

File: lightning/src/chain/channelmonitor.rs [hunk 5]
```diff
@@ -869,6 +869,9 @@ impl<Signer: Sign> Writeable for ChannelMonitorImpl<Signer> {
 			writer.write_all(&txid[..])?;
 			writer.write_all(&byte_utils::be64_to_array(htlc_infos.len() as u64))?;
 			for &(ref htlc_output, ref htlc_source) in htlc_infos.iter() {
+				debug_assert!(htlc_source.is_none() || Some(**txid) == self.current_counterparty_commitment_txid
+						|| Some(**txid) == self.prev_counterparty_commitment_txid,
+					"HTLC Sources for all but revoked commitment transactions should be none!");
```
Comment (valentinewallace) [inline lightning/src/chain/channelmonitor.rs:874]: s/but //?

Comment (valentinewallace) [reply lightning/src/chain/channelmonitor.rs:874]: Yeah, this says the opposite -- it says sources for *non*-revoked should be None

**Review by tnull — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]
